Use new admin users table to authentication and authorization.

garrettqmartin8 opened the pull request:
## Description

We currently [can't log into](https://docs.google.com/document/d/1Sb6S5bgsqyvfyxJyqLusD5s3n4ZmNDSjdPkEeMrCFYI/edit#) the admin app on dev, experimental, or staging. We're also [not properly authorizing users](https://www.pivotaltracker.com/n/projects/2136866/stories/166656598) for hitting admin app endpoints. The changes in this PR take advantage of the new `admin_users` table to make use of a strategy very similar to what we use to authenticate office users, both unknown (logging in for the first time) and known.

## Reviewer Notes

- How do we feel about the `AdminAuthMiddleware`?
- I've created a secure migration for bootstrapping us with a few system admin users:
  - Risa (the new product person on this feature)
  - Alexi (the original product person on this feature)
  - Me
  - Shall I go ahead and add more? The purpose of this is so we can begin verifying admin stories in staging.

## Setup

- In `local_migrations/20190710174257_add_system_admin_users.sql`, edit one of the insert statements to reflect the email address you use with login.gov.
- `make db_dev_e2e_migrate`
- `make server_run`
- `make client_run`

## Code Review Verification Steps

- Visit `adminlocal:3000`
- Log in via login.gov with the email address you entered in the local migration
- Ensure you don't see an unauthorized page
- Log out
- Login in as an office user, either devlocal or via login.gov
- Attempt to visit `officelocal:3000/admin/v1/office_users` and ensure you hit an `Unauthorized` page


* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/167018198) for this change

garrettqmartin8 commented on pkg/auth/authentication/auth.go:
```diff
@@ -70,6 +71,26 @@ func UserAuthMiddleware(logger Logger) func(next http.Handler) http.Handler {
+			session := auth.SessionFromRequestContext(r)
+
+			if session == nil || !session.IsAdminUser() {
+				http.Error(w, http.StatusText(403), http.StatusForbidden)
```
I first used a `401` to mirror what was in the `UseAuthMiddleware`, but realized that, confusingly, `401 Unauthorized` is actually meant for _authentication_, not authorization. Here's a [Stackoveflow post](https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses) for reference. Does anyone have strong opinions on this?

chrisgilmerproj replied:
403 seems fine to me.

chrisgilmerproj commented on local_migrations/20190710174257_add_system_admin_users.sql:
```diff
@@ -0,0 +1,21 @@
+-- intend to apply on production, but do not include any sensitive data.
+
+INSERT INTO organizations (id, name, created_at, updated_at)
+VALUES (uuid_generate_v4(), 'Truss', now(), now());
```
Let's not generate the UUID. I believe there was a decision a while ago never to use `uuid_generate_v4()`. You can `brew install uuidgen` if you need something to give you these.  Same goes for any secure migrations.

chrisgilmerproj commented on local_migrations/20190710174257_add_system_admin_users.sql:
```diff
@@ -0,0 +1,21 @@
+-- This will be run on development environments. It should mirror what you
+-- intend to apply on production, but do not include any sensitive data.
+
+INSERT INTO organizations (id, name, created_at, updated_at)
```
For completeness I'd love if we could also add `poc_email` and `poc_phone`. I'm not sure they're allowed to be NULL anyway.

garrettqmartin8 replied:
Who do you think we should we designate as the point of contact?

chrisgilmerproj replied:
It can be anything, just ensure that it has data. It's for local migrations only so just add something to help others know what to expect if they use the table.

chrisgilmerproj commented on migrations/20190708201347_make_admin_users_fname_and_lname_required.up.fizz:
```diff
@@ -0,0 +1,2 @@
+change_column("admin_users", "first_name", "string", {})
+change_column("admin_users", "last_name", "string", {})
```
I think this just changes them from varchar(255) to text? If you want to make then required I think you probably need to explicitly alter the table.

garrettqmartin8 replied:
I tried putting `"null": false` in the options, but it doesn't look like it added a `NOT NULL` to the schema. Is there another way I should approach this?

chrisgilmerproj replied:
From https://www.postgresql.org/docs/10/sql-altertable.html I'd use 

```sql
ALTER TABLE admin_users ALTER COLUMN first_name SET NOT NULL;
```

garrettqmartin8 replied:
Gotcha. Weird that Pop/Fizz doesn't have an easy way to do that without dropping down to SQL.

chrisgilmerproj commented on pkg/auth/authentication/auth.go:
```diff
@@ -50,7 +50,8 @@ func UserAuthMiddleware(logger Logger) func(next http.Handler) http.Handler {
 				return
-			} else if session.IsAdminApp() && !session.IsSuperuser {
+			} else if session.IsAdminApp() &&
+				!session.IsAdminUser() {
```
Why not `session.AdminUserID == uuid.Nil` instead of `!session.IsAdminUser()`? Alternatively, why not change the other ones above this to use the same format of `!session.IsOfficeUser()` etc?

garrettqmartin8 replied:
Good question. I like using method on session instead of the comparison so we can abstract the concern a bit. I'll convert the other ones above to use the session methods.

chrisgilmerproj commented on pkg/auth/authentication/auth.go:
```diff
@@ -70,6 +71,26 @@ func UserAuthMiddleware(logger Logger) func(next http.Handler) http.Handler {
+func AdminAuthMiddleware(logger Logger) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		mw := func(w http.ResponseWriter, r *http.Request) {
+			ctx, span := beeline.StartSpan(r.Context(), "UserAuthMiddleware")
```
```suggestion
			ctx, span := beeline.StartSpan(r.Context(), "AdminAuthMiddleware")
```

chrisgilmerproj commented on cmd/milmove/serve.go:
```diff
@@ -767,6 +767,7 @@ func serveFunction(cmd *cobra.Command, args []string) error {
 	adminAPIMux := goji.SubMux()
 	adminMux.Handle(pat.New("/*"), adminAPIMux)
 	adminAPIMux.Use(userAuthMiddleware)
+	adminAPIMux.Use(authentication.AdminAuthMiddleware(logger))
```
Is the intent here to block this API route based on your session?

garrettqmartin8 replied:
Yep.

chrisgilmerproj commented on migrations/20190710174257_add_system_admin_users.up.fizz:
```diff
@@ -0,0 +1 @@
+exec("./apply-secure-migration.sh 20190710174257_add_system_admin_users.sql")
```
I see you adding new users with inserts in this migration. But its not everyone, just you, Alexi, and Risa.  Why not read the `users` table, grab everyone who is already marked as `is_superuser` and then fill the table that way? Alternatively, can we just add all Truss personnel to Experimental and Staging and add a limited set to Prod? Not sure who you want on Prod but maybe its everyone as well.

garrettqmartin8 replied:
My [previous PR](https://github.com/transcom/mymove/pull/2346) actually did take everyone from the users table and create an admin user record for them - my question would be, do we know if anyone was actually marked as a superuser? I thought the case would be "no", but if you know differently I would love to discuss.

chrisgilmerproj replied:
Yeah, several folks were marked. You can find out who if you download all the secure migrations and look at them.

garrettqmartin8 replied:
Gotcha, I'll do that.

garrettqmartin8 replied:
Welp, it looks like a bunch of people were marked as superusers. So this secure migration only needs to include Risa. Everyone else should already have an admin_user record in the database as a result of the migration in the PR I linked above. I will change the secure migration to reflect that.

chrisgilmerproj commented on pkg/auth/authentication/devlocal.go:
```diff
@@ -75,6 +75,7 @@ func (h UserListHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		DpsUserType     string
 		IsAdminApp      bool
 		AdminUserType   string
+		IsAdminUser     bool
```
Can't we just use `IsAdminApp` here?

garrettqmartin8 replied:
Ah, yeah. This is vestigial from me throwing stuff against the wall early on.

mikena-truss commented on pkg/auth/authentication/auth.go:
```diff
@@ -418,6 +436,41 @@ func authorizeKnownUser(userIdentity *models.UserIdentity, h CallbackHandler, se
+			session.AdminUserRole = adminUser.Role.String()
+			span.AddField("session.admin_user_id", session.AdminUserID)
+			adminUser.UserID = &userIdentity.ID
+			err = h.db.Save(adminUser)
```
Any reason not to use `ValidateAndSave` here?

garrettqmartin8 replied:
Nope! I'll change it.

mikena-truss commented on pkg/auth/session.go:
```diff
@@ -102,7 +103,19 @@ func (s *Session) IsTspUser() bool {
+
+// IsSystemAdmin checks whether the authenticated admin user is a system admin
+func (s *Session) IsSystemAdmin() bool {
+	role := "SYSTEM_ADMIN"
```
We might want to use a similar to pattern to how we do move states here.

In the model:
```
type AdminRole

const (
	SystemAdminRole AdminRole = "SYSTEM_ADMIN"
	ProgramAdminRole AdminRole = "PROGRAM_ADMIN"
)
```

Then here you get `models.SystemAdminRole` rather than redefining the constants.

mikena-truss replied:
Right now it looks like we use a `ValidRoles` function, but I think that can work hand in hand we these `const`.

mikena-truss commented on pkg/models/admin_user.go:
```diff
@@ -69,3 +76,15 @@ func (a *AdminUser) ValidateCreate(tx *pop.Connection) (*validate.Errors, error)
 	return validate.NewErrors(), nil
 }
+
+func FetchAdminUserByEmail(tx *pop.Connection, email string) (*AdminUser, error) {
```
Any chance we could just use the `FetchOne` here? Part of creating that method was to get rid of the pop package leaking into the models. It might be a good candidate here. Might have to get creative with the `LOWER` postgres function though...

garrettqmartin8 replied:
Trying to figure out a way to use `FetchOne`, but using the `LOWER` function is proving to be difficult without changing the query builder. It looks like the user model is converting email to lowercase on creation:

```go
func CreateUser(db *pop.Connection, loginGovID string, email string) (*User, error) {
     lgu, err := uuid.FromString(loginGovID)
     if err != nil {
         return nil, err
     }
     newUser := User{
         LoginGovUUID:  lgu,
         LoginGovEmail: strings.ToLower(email),
         Disabled:      false,
     }
     verrs, err := db.ValidateAndCreate(&newUser)
     if verrs.HasAny() {
         return nil, verrs
     } else if err != nil {
         err = errors.Wrap(err, "Unable to create user")
         return nil, err
     }
     return &newUser, nil
 }
```

Given that, can we safely compare the email without any sql tranformation function? I have all the tests passing when I do this:

```go
var adminUser models.AdminUser
queryBuilder := query.NewQueryBuilder(h.db)
filters := []services.QueryFilter{
                 query.NewQueryFilter("email", "=", strings.ToLower(userIdentity.Email)),
}
err := queryBuilder.FetchOne(&adminUser, filters)
```

mikena-truss replied:
Thanks for poking around with this. There might be a backwards compatibility issue here with existing emails. If this is a pain, it's not a huge deal with use the builder, I just thought it would be nice. It's more important to use it for the response data. I'll leave it up to you.

garrettqmartin8 replied:
If we're downcasing emails when users are created, would this actually create a backwards compatibility issue? My understanding is that we are indeed making sure emails are downcased when users are created. I'm cool with reverting these changes if I'm missing some piece of context here.

mikena-truss replied:
Oh, I misread that and missed that `CreateUser` `ToLower` was existing code. 👍

mikena-truss commented on pkg/models/admin_user.go:
```diff
@@ -69,3 +76,15 @@ func (a *AdminUser) ValidateCreate(tx *pop.Connection) (*validate.Errors, error)
+	if len(users) == 0 {
+		return nil, ErrFetchNotFound
+	}
+	return &users[0], nil
```
Any reason not to use `First` instead of `All` and `[0]`?

garrettqmartin8 commented:
@mikena-truss I've made some changes based on your feedback.

garrettqmartin8 commented:
@chrisgilmerproj would you mind taking another look? Still one outstanding questions about point of contact.

Ryan-Koch commented on pkg/auth/authentication/auth.go:
```diff
@@ -418,6 +438,52 @@ func authorizeKnownUser(userIdentity *models.UserIdentity, h CallbackHandler, se
+			err := queryBuilder.FetchOne(&adminUser, filters)
+			if err == models.ErrFetchNotFound {
+				h.logger.Error("Non-admin user authenticated at admin site", zap.String("email", session.Email))
+				http.Error(w, http.StatusText(401), http.StatusUnauthorized)
```
Would we want to return a 404 or a 403 in this case? At it's most literal, a 401 invites a repeat attempt at the request. Is this a case where we'd want that behavior potentially or does this instance of not found fall into a category with more permanence?

garrettqmartin8 replied:
I think a 403 probably makes sense here.

mikena-truss commented on pkg/models/admin_user.go:
```diff
@@ -12,24 +13,35 @@ import (
-		"SYSTEM_ADMIN",
-		"PROGRAM_ADMIN",
+const (
+	SystemAdminRole  AdminRole = "SYSTEM_ADMIN"
```
It looks like this got added but isn't being used outside of this package. Anywhere `"SYSTEM_ADMIN"` is used it should be `models.SystemAdminRole`, unless it would create a weird package import.

garrettqmartin8 replied:
Yeah, it creates a package import issue when I try to use `models.SystemAdminRole` within the `auth` package:
```
package github.com/transcom/mymove/cmd/generate-test-data
	imports github.com/transcom/mymove/pkg/models
	imports github.com/transcom/mymove/pkg/auth
	imports github.com/transcom/mymove/pkg/models
import cycle not allowed
package github.com/transcom/mymove/internal/pkg/dutystationsloader
	imports github.com/transcom/mymove/pkg/models
	imports github.com/transcom/mymove/pkg/auth
	imports github.com/transcom/mymove/pkg/models
```

garrettqmartin8 replied:
But I would really like to be able to use the constant instead of the raw string. Hmmm...

mikena-truss replied:
Sounds like it's alluding to role based auth where that info doesn't necessarily come from the models 😭

Anyways, I think that's ok for now 👍

mikena-truss commented:
Left a couple minor comments. Other than that looks good.

garrettqmartin8 commented:
@chrisgilmerproj I've addressed 1 & 3 and in an updated secure migration.

Regarding number 2: I've replaced as many references to superuser as I've found and have been rebasing master frequently to make sure I'm not letting any slip through the cracks. Are there other cases that I'm not thinking of?

chrisgilmerproj commented on migrations/20190709211634_remove_is_superuser.up.fizz:
```diff
@@ -0,0 +1 @@
+drop_column("users", "is_superuser")
```
@garrettqmartin8 - My suggestion is that you do this migration in a totally different PR that lands after you deploy this branch. Migrations are always deployed before code so this will effectively drop the column and bork the running server before your code goes out, which could lead to a service interruption on the order of 10ish minutes. Once you pull this out of the PR I'll be happy to approve so you can merge. And we can look at getting this migration in on Tuesday or Wednesday.

garrettqmartin8 replied:
Got it - I forgot that migrations were deployed before code. Updated!
